### PR TITLE
Make definitions block-level

### PR DESCRIPTION
**Pull request by talex5**
I just converted my blog from Octopress to omd, and I needed support for definition lists with multiple paragraphs.

Not completely sure what I'm doing here, but this changed got it working it for me, and should go some way towards #221.

### CODE REVIEW

File: src/omd.mli [hunk 1]
```diff
@@ -30,18 +30,18 @@ and 'attr inline =
...
+and 'attr block =
   | Paragraph of 'attr * 'attr inline
   | List of 'attr * list_type * list_spacing * 'attr block list list
   | Blockquote of 'attr * 'attr block list
   | Thematic_break of 'attr
   | Heading of 'attr * int * 'attr inline
   | Code_block of 'attr * string * string
   | Html_block of 'attr * string
-  | Definition_list of 'attr * 'attr def_elt list
+  | Definition_list of 'attr * list_spacing * 'attr def_elt list
```
Comment (shonfeder) [inline src/omd.mli:44]: If this is to become a recursive type, perhaps the `block` should be listed first, since it's the "principle" type?

File: tests/def_list.md [hunk 2]
```diff
@@ -10,13 +10,80 @@ Second Term
...
+# With multiple paragraphs
+
+First Term
+: This is the definition of the first term.
+
+Second Term
+: This is one paragraph of the second term.
+
+  This is another paragraph of the second term.
+
```
Comment (shonfeder) [inline tests/def_list.md:23]: Perhaps we should include cases for

- mixing lazy multiline and the multi-paragraph markup
- mixing using multiple definition terms with the multi-paragraph markup
- some case like

   ```
   # foo
   : not a definition
   ```

**Review by shonfeder — COMMENTED** [review]
Thanks for the improvement! This will definitely be useful.

I've a few suggestion, but I'd be happy to make them on this PR if you're not inclined to. Just let me know if you'd prefer that.

**Comment by shonfeder** [comment]
Looks like we'll also need to run ocamlformat, btw.

**Comment by talex5** [comment]
> I've a few suggestion, but I'd be happy to make them on this PR if you're not inclined to. Just let me know if you'd prefer that.

That would be great - thanks! Your suggestions all sound sensible to me.

File: tests/def_list.md [hunk 3]
```diff
@@ -10,13 +10,122 @@ Second Term
...
+<p>This is one paragraph using lazy wrapping,
+so this should be part of the first paragraph
+of the third term.</p>
+<p>But this should be the second paragraph of
+the second paragraph</p>
+<p>Note that, currently, subsequent paragraphs cannot lazy wrap.</p>
+</dd>
+</dl>
+<p>For example, this line will be part of a new paragraph outside
+of the definition list entirely.</p>
```
Comment (shonfeder) [inline tests/def_list.md:96]: @talex5 does this seem like the right behavior to you? I'm not exactly sure what ought to happen in this case. I can report that in pandoc, this markdown

```
Term 1
: A first paragraph that is
lazily wrapped

    A second paragrap that 
is lazily wrapped 
```

is rendered as 

```html
<dl>
<dt>Term 1</dt>
<dd><p>A first paragraph that is lazily wrapped</p>
<p>A second paragrap that is lazily wrapped</p>
</dd>
</dl>
```

But they use 4 spaces to indicate the start of the second paragraph.

**Review by shonfeder — COMMENTED** [review]
Pushed my changes up. I do have one pending question for you about some edge case behavior after I added a few more things to the test cases.

File: src/block.ml [hunk 4]
```diff
@@ -147,10 +181,34 @@ module Pre = struct
...
+            Rdef_list
+              { deflist with
+                empty_line_seen = true
+              ; state = process deflist.state s
+              }
         }
+    | Rdef_list deflist, _ when Parser.indent s >= deflist.indent ->
+        let s = Sub.offset deflist.indent s in
+        let state = process deflist.state s in
+        { blocks; next = Rdef_list { deflist with state } }
```
Comment (talex5) [inline src/block.ml:207]: I guess this should set `empty_line_seen = false` to fix the problem with lazy wrapping you mention below.

File: src/block.ml [hunk 5]
```diff
@@ -147,10 +181,34 @@ module Pre = struct
         { blocks
         ; next = Rfenced_code (ind, num, q, info, Sub.to_string s :: lines, a)
         }
-    | Rdef_list (term, d :: defs), Lparagraph ->
+    | ( Rdef_list ({ empty_line_seen = false; _ } as deflist)
```
Comment (talex5) [inline src/block.ml:184]: Thinking about this a bit more, I don't think we need the `empty_line_seen = false` check here.

File: src/block.ml (hunk 4 above)
Comment (shonfeder) [reply src/block.ml:207]: This does address the lazy wrapping, thanks for the tip! Unfortunately it also breaks the nested def-list parsing. I'll continue considering it, but if an immediate solution occurs to you, let me know ;)

File: src/block.ml (hunk 5 above)
Comment (shonfeder) [reply src/block.ml:184]: Changing this (either by removing the match on the `empty_line_seen` or removing the case all together) leads to failures around the `This is not a correct definition list` line. This to will need some considering. The moral of the story seems to me that the handwritten parsing logic here is quite touchy and tough to reason about...

Comment (talex5) [reply src/block.ml:184]: https://github.com/jgm/commonmark-hs/blob/master/commonmark-extensions/test/definition_lists.md ("Multiple definitions, loose") gives this example:

```
apple

:   red fruit

:   computer company

orange

:   orange fruit
:   telecom company
.
<dl>
<dt>apple</dt>
<dd>
<p>red fruit</p>
</dd>
<dd>
<p>computer company</p>
</dd>
<dt>orange</dt>
<dd>
<p>orange fruit</p>
</dd>
<dd>
<p>telecom company</p>
</dd>
</dl>
```
So perhaps the `This is not a correct definition list` test is itself wrong?

Comment (shonfeder) [reply src/block.ml:184]: Ah, thanks for the reference! I'm happy with deferring to commonmark on these things. I'll spend some time comparing their test suite with ours later this week or over the weekend.

Comment (shonfeder) [reply src/block.ml:184]: Ah, duh. Now that I look at the test case, since you've introduced the loose spacing, of course that test is now incorrect! My bad for not giving it sufficient thought before.

Comment (shonfeder) [reply src/block.ml:184]: So, if we remove this check, then it fixes the loose format for multiple definitions, but it breaks the nesting. Which I think means that, even if we were OK with the leaving the lazy line wrapping broken for multi-paragraph definitions, we're still a bit stuck on this point.

I'll investigate over the weekend, and try to get my head around properly around the parsing function. But just an FYI to explains why I'm not merging this straight away.

File: src/block.ml (hunk 4 above)
Comment (sonologico) [reply src/block.ml:207]: Reordering the clauses such that the one that checks for the indentation level comes first solves the problem with the nested def-list test. The only remaining problem on def_list-003 test then are missing `<p>`s in the generated html.
